File: AlDat/Ueb5.py
def getMedian(C: list, CIndex: int, n: int) -> float:
    if (n % 2 == 0):
        return (C[CIndex + n//2 - 1] + C[CIndex + n//2])/2
    else:
        return C[CIndex + n//2]


def median(A: list, AIndex: int, B: list, BIndex: int, n: int) -> float:
    if(n == 0):
        return 0
    if(n == 1):
        return (A[AIndex] + B[BIndex])/2
    if(n == 2):
        C = [A[AIndex], A[AIndex+1], B[BIndex], B[BIndex+1]]
        C.sort()
        return getMedian(C,0, 4)

    m1 = getMedian(A, AIndex, n)
    m2 = getMedian(B, BIndex, n)

    if(m1 == m2):
        return m1
    if (m1 < m2):
        return median(A, AIndex+n//2, B, BIndex, n - n//2)
    else:
        return median(A, AIndex, B, BIndex + n//2, n - n//2)


A = [1,3,6,12,14]
B = [2,5,7,8,10]
print(median(A, 0, B, 0, len(A)))

File: AlDat/Ueb5.c
#include <stdio.h>
#include <stdlib.h>

double getMedian(int C[], int n) // O(1)
{
    if (n % 2 == 0) // gerade Anzahl an Elementen => Durchschnitt der Mitte
    {
        return (C[n / 2] + C[n / 2 - 1]) / 2.0;
    }
    else // ungerade Anzahl => mittleres Element
    {
        return C[n / 2];
    }
}

int cmpfunc(const void *a, const void *b) // O(1)
{
    return (*(int *)a - *(int *)b);
}

double median(int A[], int B[], int n) // O(log n)
{
    if (n == 0) // Leeres Array => kein Element
    {
        return 0; // O(1)
    }
    else if (n == 1) // Ein Element in jedem Array => Durchschnitt der Beiden
    {
        return (A[0] + B[0]) / 2.0; // O(1)
    }
    else if (n == 2) // Zwei Elemente in jedem Array, Median = Median der 4 Elemente
    {
        int C[] = {A[0], A[1], B[0], B[1]}; // O(1)
        qsort(C, 4, sizeof(int), cmpfunc);  // O(4 log 4) = O(1)
        return getMedian(C, 4);             // O(1)
    }

    double mA = getMedian(A, n); // O(1)
    double mB = getMedian(B, n); // O(1)

    if (mA == mB) // mA <= m <= mB => m = mA = mB
    {
        return mA; // O(1)
    }
    else if (mA < mB) // mA < m < mB => suchen auf der oberen Hälfte von A und unteren von B
    {
        return median(A + n / 2, B, n - n / 2); // log n Aufrufe
    }
    else // mA > m > mB => suchen auf der oberen Hälfte von B und unteren von A
    {
        return median(A, B + n / 2, n - n / 2); // log n Aufrufe
    }
}

int main(int argc, char const *argv[])
{
    int A[] = {1, 5, 6, 7, 8};
    int B[] = {2, 3, 4, 9, 10};
    printf("%lf\n", median(A, B, 5));
    return 0;
}
